File: backend/src/connectors/registry.ts
import type { ConnectorProvider } from "./plugin-types";

const providerRegistry = new Map<string, ConnectorProvider>();

export function registerConnectorProvider(provider: ConnectorProvider): ConnectorProvider {
  providerRegistry.set(provider.manifest.key, provider);
  return provider;
}

export function getConnectorProvider(kind: string): ConnectorProvider {
  const registered = providerRegistry.get(kind);
  if (registered) return registered;
  throw new Error(`No connector provider registered for kind: ${kind}`);
}

export async function loadConnectors() {
  await Promise.all([
    import("./google-drive")
  ]);
}

File: backend/app/connectors/registry.py
from __future__ import annotations

from app.connectors.plugin_types import ConnectorProvider

_registry: dict[str, ConnectorProvider] = {}


def register_connector_provider(kind: str, provider: ConnectorProvider) -> None:
    _registry[kind] = provider


def get_connector_provider(kind: str) -> ConnectorProvider:
    if kind not in _registry:
        raise KeyError(f"Unknown connector provider: {kind}")
    return _registry[kind]


def load_connectors() -> None:
    import app.connectors.google_drive.index  # noqa: F401
    import app.connectors.slack.index  # noqa: F401
